In [1]:
import pandas as pd
import os

In [3]:
os.listdir()

['.ipynb_checkpoints',
 '01_data_inspection.ipynb',
 'Data Sprint Guidelines (Olist) .pdf',
 'olist_closed_deals_dataset.csv',
 'olist_customers_dataset.csv.zip',
 'Olist_Data_Sprint_Project_Manager_Guide.docx',
 'olist_geolocation_dataset.csv.zip',
 'olist_marketing_qualified_leads_dataset.csv',
 'olist_orders_dataset.csv.zip',
 'olist_order_items_dataset.csv.zip',
 'olist_order_payments_dataset.csv.zip',
 'olist_order_reviews_dataset.csv.zip',
 'olist_products_dataset.csv.zip',
 'olist_sellers_dataset.csv',
 'Olist_Workflow_Project_Plan_Tracker.xlsx',
 'product_category_name_translation.csv',
 '~$Olist_Workflow_Project_Plan_Tracker.xlsx']

In [5]:
mql = pd.read_csv("olist_marketing_qualified_leads_dataset.csv")
closed = pd.read_csv("olist_closed_deals_dataset.csv")

In [7]:
print("MQL shape:", mql.shape)
print("Closed Deals shape:", closed.shape)

display(mql.head())
display(closed.head())

MQL shape: (8000, 4)
Closed Deals shape: (842, 14)


,mql_id,first_contact_date,landing_page_id,origin
0,dac32acd4db4c29c230538b72f8dd87d,2018-02-01,88740e65d5d6b056e0cda098e1ea6313,social
1,8c18d1de7f67e60dbd64e3c07d7e9d5d,2017-10-20,007f9098284a86ee80ddeb25d53e0af8,paid_search
2,b4bc852d233dfefc5131f593b538befa,2018-03-22,a7982125ff7aa3b2054c6e44f9d28522,organic_search
3,6be030b81c75970747525b843c1ef4f8,2018-01-22,d45d558f0daeecf3cccdffe3c59684aa,email
4,5420aad7fec3549a85876ba1c529bd84,2018-02-21,b48ec5f3b04e9068441002a19df93c6c,organic_search


,mql_id,seller_id,sdr_id,sr_id,won_date,business_segment,lead_type,lead_behaviour_profile,has_company,has_gtin,average_stock,business_type,declared_product_catalog_size,declared_monthly_revenue
0,5420aad7fec3549a85876ba1c529bd84,2c43fb513632d29b3b58df74816f1b06,a8387c01a09e99ce014107505b92388c,4ef15afb4b2723d8f3d81e51ec7afefe,2018-02-26 19:58:54,pet,online_medium,cat,NaN,NaN,NaN,reseller,NaN,0.0
1,a555fb36b9368110ede0f043dfc3b9a0,bbb7d7893a450660432ea6652310ebb7,09285259593c61296eef10c734121d5b,d3d1e91a157ea7f90548eef82f1955e3,2018-05-08 20:17:59,car_accessories,industry,eagle,NaN,NaN,NaN,reseller,NaN,0.0
2,327174d3648a2d047e8940d7d15204ca,612170e34b97004b3ba37eae81836b4c,b90f87164b5f8c2cfa5c8572834dbe3f,6565aa9ce3178a5caf6171827af3a9ba,2018-06-05 17:27:23,home_appliances,online_big,cat,NaN,NaN,NaN,reseller,NaN,0.0
3,f5fee8f7da74f4887f5bcae2bafb6dd6,21e1781e36faf92725dde4730a88ca0f,56bf83c4bb35763a51c2baab501b4c67,d3d1e91a157ea7f90548eef82f1955e3,2018-01-17 13:51:03,food_drink,online_small,NaN,NaN,NaN,NaN,reseller,NaN,0.0
4,ffe640179b554e295c167a2f6be528e0,ed8cb7b190ceb6067227478e48cf8dde,4b339f9567d060bcea4f5136b9f5949e,d3d1e91a157ea7f90548eef82f1955e3,2018-07-03 20:17:45,home_appliances,industry,wolf,NaN,NaN,NaN,manufacturer,NaN,0.0


In [9]:
print("MQL columns:")
print(mql.columns.tolist())

print("\nClosed Deals columns:")
print(closed.columns.tolist())

MQL columns:
['mql_id', 'first_contact_date', 'landing_page_id', 'origin']

Closed Deals columns:
['mql_id', 'seller_id', 'sdr_id', 'sr_id', 'won_date', 'business_segment', 'lead_type', 'lead_behaviour_profile', 'has_company', 'has_gtin', 'average_stock', 'business_type', 'declared_product_catalog_size', 'declared_monthly_revenue']


In [11]:
mql.info()
print("\n" + "="*60 + "\n")
closed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   mql_id              8000 non-null   object
 1   first_contact_date  8000 non-null   object
 2   landing_page_id     8000 non-null   object
 3   origin              7940 non-null   object
dtypes: object(4)
memory usage: 250.1+ KB


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   mql_id                         842 non-null    object 
 1   seller_id                      842 non-null    object 
 2   sdr_id                         842 non-null    object 
 3   sr_id                          842 non-null    object 
 4   won_date                       842 non-null    object 
 5   business_segment               8

In [13]:
print("MQL missing values:")
display(mql.isna().sum().sort_values(ascending=False))

print("\nClosed Deals missing values:")
display(closed.isna().sum().sort_values(ascending=False))

MQL missing values:


origin                60
mql_id                 0
first_contact_date     0
landing_page_id        0
dtype: int64


Closed Deals missing values:


has_company                      779
has_gtin                         778
average_stock                    776
declared_product_catalog_size    773
lead_behaviour_profile           177
business_type                     10
lead_type                          6
business_segment                   1
mql_id                             0
seller_id                          0
sdr_id                             0
sr_id                              0
won_date                           0
declared_monthly_revenue           0
dtype: int64

In [15]:
print("MQL duplicate mql_id:", mql["mql_id"].duplicated().sum())
print("Closed duplicate mql_id:", closed["mql_id"].duplicated().sum())
print("Closed duplicate seller_id:", closed["seller_id"].duplicated().sum())

MQL duplicate mql_id: 0
Closed duplicate mql_id: 0
Closed duplicate seller_id: 0


In [17]:
print("Acquisition channels:")
display(mql["origin"].value_counts(dropna=False))

print("\nBusiness types:")
display(closed["business_type"].value_counts(dropna=False))

print("\nLead types:")
display(closed["lead_type"].value_counts(dropna=False))

print("\nBusiness segments:")
display(closed["business_segment"].value_counts(dropna=False))

Acquisition channels:


origin
organic_search       2296
paid_search          1586
social               1350
unknown              1099
direct_traffic        499
email                 493
referral              284
other                 150
display               118
other_publicities      65
NaN                    60
Name: count, dtype: int64


Business types:


business_type
reseller        587
manufacturer    242
NaN              10
other             3
Name: count, dtype: int64


Lead types:


lead_type
online_medium      332
online_big         126
industry           123
offline            104
online_small        77
online_beginner     57
online_top          14
NaN                  6
other                3
Name: count, dtype: int64


Business segments:


business_segment
home_decor                         105
health_beauty                       93
car_accessories                     77
household_utilities                 71
construction_tools_house_garden     69
audio_video_electronics             64
computers                           34
pet                                 30
food_supplement                     28
food_drink                          26
sports_leisure                      25
bags_backpacks                      22
bed_bath_table                      22
toys                                20
fashion_accessories                 19
home_office_furniture               14
phone_mobile                        13
stationery                          13
small_appliances                    12
handcrafted                         12
baby                                10
music_instruments                    9
books                                9
watches                              8
jewerly                              8
home_app

In [19]:
print("Unique SDRs:", closed["sdr_id"].nunique())
print("Unique SRs:", closed["sr_id"].nunique())
print("Unique sellers:", closed["seller_id"].nunique())

Unique SDRs: 32
Unique SRs: 22
Unique sellers: 842


# INITIAL EXPLORATORY ANALYSIS
# Preliminary observations only — before cleaning and validation

In [22]:
funnel_initial = mql.merge(closed,on="mql_id",how="left")

In [28]:
print(mql.shape)
print(closed.shape)
print(funnel_initial.shape)

(8000, 4)
(842, 14)
(8000, 17)


In [30]:
funnel_initial["converted"] = funnel_initial["seller_id"].notna()

In [32]:
funnel_initial["converted"].value_counts()

converted
False    7158
True      842
Name: count, dtype: int64

In [34]:
conversion_rate_initial = funnel_initial["converted"].mean() * 100

In [36]:
conversion_rate_initial

10.525

In [38]:
print("Initial conversion rate:", round(conversion_rate_initial, 2), "%")

Initial conversion rate: 10.52 %


In [ ]:
#Which marketing channels seem to convert leads into sellers more often?


In [40]:
mql["origin"].value_counts()

origin
organic_search       2296
paid_search          1586
social               1350
unknown              1099
direct_traffic        499
email                 493
referral              284
other                 150
display               118
other_publicities      65
Name: count, dtype: int64

In [42]:
funnel_initial[funnel_initial["converted"] == True]["origin"].value_counts()

origin
organic_search       271
paid_search          195
unknown              179
social                75
direct_traffic        56
referral              24
email                 15
display                6
other                  4
other_publicities      3
Name: count, dtype: int64

In [46]:
total_leads_by_channel = mql["origin"].value_counts()

converted_by_channel = funnel_initial[funnel_initial["converted"] == True]["origin"].value_counts()

In [48]:
channel_conversion_rate = (converted_by_channel / total_leads_by_channel) * 100

In [50]:
channel_conversion_rate.sort_values(ascending=False)

origin
unknown              16.287534
paid_search          12.295082
organic_search       11.803136
direct_traffic       11.222445
referral              8.450704
social                5.555556
display               5.084746
other_publicities     4.615385
email                 3.042596
other                 2.666667
Name: count, dtype: float64

Preliminary observation: Lead-to-seller conversion differs across acquisition channels. paid_search, organic_search, and direct_traffic currently show conversion rates above the overall preliminary rate of 10.52%. The unknown category has the highest observed rate, but it cannot yet be interpreted as an actionable acquisition channel. These findings are exploratory and will be validated after data cleaning and seller-performance analysis.

In [53]:
orders = pd.read_csv("olist_orders_dataset.csv.zip")
order_items = pd.read_csv("olist_order_items_dataset.csv.zip")
sellers = pd.read_csv("olist_sellers_dataset.csv")
reviews = pd.read_csv("olist_order_reviews_dataset.csv.zip")

In [55]:
print("Orders:", orders.shape)
print("Order Items:", order_items.shape)
print("Sellers:", sellers.shape)
print("Reviews:", reviews.shape)

Orders: (99441, 8)
Order Items: (112650, 7)
Sellers: (3095, 4)
Reviews: (99224, 7)


In [57]:
print("Orders columns:")
print(orders.columns.tolist())

Orders columns:
['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp', 'order_approved_at', 'order_delivered_carrier_date', 'order_delivered_customer_date', 'order_estimated_delivery_date']


In [59]:
print("\nOrder Items columns:")
print(order_items.columns.tolist())


Order Items columns:
['order_id', 'order_item_id', 'product_id', 'seller_id', 'shipping_limit_date', 'price', 'freight_value']


In [61]:
print("\nSellers columns:")
print(sellers.columns.tolist())


Sellers columns:
['seller_id', 'seller_zip_code_prefix', 'seller_city', 'seller_state']


In [63]:
print("\nReviews columns:")
print(reviews.columns.tolist())


Reviews columns:
['review_id', 'order_id', 'review_score', 'review_comment_title', 'review_comment_message', 'review_creation_date', 'review_answer_timestamp']


In [65]:
display(orders.head())

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [67]:
display(order_items.head())

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [69]:
display(sellers.head())

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [71]:
display(reviews.head())

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [73]:
customers = pd.read_csv("olist_customers_dataset.csv.zip")

print("Customers shape:", customers.shape)
print(customers.columns.tolist())

display(customers.head())

Customers shape: (99441, 5)
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [75]:
print("Orders duplicate order_id:",
      orders["order_id"].duplicated().sum())

Orders duplicate order_id: 0


In [77]:
print("Order Items duplicate order_id:",
      order_items["order_id"].duplicated().sum())

Order Items duplicate order_id: 13984


In [79]:
print("Sellers duplicate seller_id:",
      sellers["seller_id"].duplicated().sum())

Sellers duplicate seller_id: 0


In [81]:
print("Customers duplicate customer_id:",
      customers["customer_id"].duplicated().sum())

Customers duplicate customer_id: 0


In [83]:
print("Reviews duplicate order_id:",
      reviews["order_id"].duplicated().sum())

Reviews duplicate order_id: 551


In [85]:
print("Unique orders:", orders["order_id"].nunique())

Unique orders: 99441


In [87]:
print("Unique sellers in sellers table:", sellers["seller_id"].nunique())

Unique sellers in sellers table: 3095


In [89]:
print("Unique sellers in order items:", order_items["seller_id"].nunique())

Unique sellers in order items: 3095


In [91]:
print("Unique customers:", customers["customer_unique_id"].nunique())

Unique customers: 96096


In [93]:
#missing values 
print("ORDERS")
print(orders.isna().sum())

print("\nORDER ITEMS")
print(order_items.isna().sum())

print("\nSELLERS")
print(sellers.isna().sum())

print("\nCUSTOMERS")
print(customers.isna().sum())

print("\nREVIEWS")
print(reviews.isna().sum())

ORDERS
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_customer_date    2965
order_estimated_delivery_date       0
dtype: int64

ORDER ITEMS
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

SELLERS
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

CUSTOMERS
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

REVIEWS
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message    

In [95]:
orders["order_status"].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

In [97]:
reviews["review_score"].value_counts().sort_index()

review_score
1    11424
2     3151
3     8179
4    19142
5    57328
Name: count, dtype: int64

In [99]:
reviews["review_score"].mean()

4.08642062404257

In [101]:
order_items["price"].describe()

count    112650.000000
mean        120.653739
std         183.633928
min           0.850000
25%          39.900000
50%          74.990000
75%         134.900000
max        6735.000000
Name: price, dtype: float64

In [103]:
order_items["freight_value"].describe()

count    112650.000000
mean         19.990320
std          15.806405
min           0.000000
25%          13.080000
50%          16.260000
75%          21.150000
max         409.680000
Name: freight_value, dtype: float64

In [105]:
sellers["seller_state"].value_counts()

seller_state
SP    1849
PR     349
MG     244
SC     190
RJ     171
RS     129
GO      40
DF      30
ES      23
BA      19
CE      13
PE       9
PB       6
RN       5
MS       5
MT       4
RO       2
SE       2
PI       1
AC       1
MA       1
AM       1
PA       1
Name: count, dtype: int64

In [107]:
sellers["seller_city"].value_counts().head(10)

seller_city
sao paulo         694
curitiba          127
rio de janeiro     96
belo horizonte     68
ribeirao preto     52
guarulhos          50
ibitinga           49
santo andre        45
campinas           41
maringa            40
Name: count, dtype: int64

In [109]:
customers["customer_state"].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

In [111]:
customers["customer_city"].value_counts().head(10)

customer_city
sao paulo                15540
rio de janeiro            6882
belo horizonte            2773
brasilia                  2131
curitiba                  1521
campinas                  1444
porto alegre              1379
salvador                  1245
guarulhos                 1189
sao bernardo do campo      938
Name: count, dtype: int64

In [113]:
orders["order_purchase_timestamp"] = pd.to_datetime(orders["order_purchase_timestamp"])

In [115]:
orders["order_delivered_customer_date"] = pd.to_datetime(orders["order_delivered_customer_date"])

In [117]:
orders["order_estimated_delivery_date"] = pd.to_datetime(orders["order_estimated_delivery_date"])

In [119]:
print("First order:",orders["order_purchase_timestamp"].min())

First order: 2016-09-04 21:15:19


In [121]:
print("Last order:",orders["order_purchase_timestamp"].max())

Last order: 2018-10-17 17:30:18


In [123]:
orders["order_purchase_timestamp"].dt.year.value_counts().sort_index()

order_purchase_timestamp
2016      329
2017    45101
2018    54011
Name: count, dtype: int64

In [125]:
orders["delivery_days_initial"] = (orders["order_delivered_customer_date"]- orders["order_purchase_timestamp"]).dt.days

In [127]:
orders["delivery_days_initial"].describe()

count    96476.000000
mean        12.094086
std          9.551746
min          0.000000
25%          6.000000
50%         10.000000
75%         15.000000
max        209.000000
Name: delivery_days_initial, dtype: float64

In [129]:
closed["seller_id"].isin(order_items["seller_id"]).value_counts()

seller_id
False    462
True     380
Name: count, dtype: int64

In [131]:
closed["seller_id"].isin(order_items["seller_id"]).mean() * 100

45.13064133016627

In [133]:
###Preliminary only — before cleaning and validation